README.md
# TYPO3 Extension intcache

[![Latest Stable Version](https://img.shields.io/packagist/v/ichhabrecht/intcache.svg)](https://packagist.org/packages/ichhabrecht/intcache)
[![Build Status](https://img.shields.io/travis/IchHabRecht/intcache/master.svg)](https://travis-ci.org/IchHabRecht/intcache)
[![StyleCI](https://github.styleci.io/repos/190002031/shield?branch=master)](https://github.styleci.io/repos/190002031)

Turn uncachable page objects into cacheable links

Pages that contain uncachable content elements (INT-objects, uncached plugin content) are delivered with a no-cache
header to the user. This extension replaces the uncachable content elements and provides urls to fetch the content
asynchronous.

The content is replaced either as
- div-container for JavaScript/Ajax processing
- SSI block for NGINX or Apache processing
- ESI block for Varnish processing

## Installation

Simply install the extension with Composer or the [Extension Manager](https://extensions.typo3.org/extension/intcache/).

`composer require ichhabrecht/intcache`

## Usage

- include the provided static TypoScript of the intcache extension

### JavaScript

- for each element a div-container with classes `intcache intcache-item intcache-link` is rendered
- the source url is provided as `data-src` attribute
- you need to provide a script that iterates over all div's and fetches the content from provided urls

### Server Side Includes

- enable SSI support in your NGINX configuration

```
location ~ \.php$ {
    ssi on;
}
```

- change TypoScript to use SSI rendering

```
lib.intcache.format = ssi
```

### Edge Side Includes

- enable ESI support in your Varnish configuration

```
sub vcl_backend_response {
    set beresp.do_esi = true;
}
```

- change TypoScript to use ESI rendering

```
lib.intcache.format = esi
```

## Additional configuration

### intcache handling

To be able to deactivate the intcache handling on certain sites and/or domains, you can explicitly disable it using
the TypoScript setup `config.intcache = 0`.

### Templates

Simply set the `templateRootPath` TypoScript *constant* to provide an additional template path. 

```
lib.intcache.view.templateRootPath = EXT:extension/Resources/Private/Templates/Intcache/
```

By default the files `Intcache.ajax`, `Intcache.esi` or `Intcache.ssi` are used for rendering (according to your current
format setting). You may want to change the `lib.intcache.format` TypoScript setting to add your own format. 

### Content

To be able the see the current content without any further processing, the content is rendered if the TYPO3 *Development*
application context is enabled.

### Cache timeout

By default all urls send a `no-cache` header to the user. You can define own cache timeouts by using `cache_timeout`
TypoScript configuration.

Example configuration for COA_INT objects:
```
page.5 = COA_INT
page.5 {
    cache_timeout = 500 // cache this content 500 seconds
    10 = TEXT
    10.wrap = <p>|</p>
    10.value = Hello world
}
```

Example configuration for plugins:
```
tt_content.list.20.[pluginName].cache_timeout = 300
```

## Community

- Thanks to [elementare teilchen GmbH](https://www.elementare-teilchen.de) that sponsors the maintenance of this extension with a [GitHub sponsorship](https://github.com/sponsors/IchHabRecht)


ChangeLog
2020-04-24         [RELEASE] Release of intcache 1.1.3 (Nicole Cordes)
2020-04-24 e2c7e25 [TASK] Raise TYPO3 dependencies to include 10.4 (Nicole Cordes)
2020-04-17 c285017 [TASK] Update community part of README.md (Nicole Cordes)

2020-04-15         [RELEASE] Release of intcache 1.1.2 (Nicole Cordes)
2020-04-14 19a6377 [BUGFIX] Add support for route enhancers (Nicole Cordes)

2020-02-10         [RELEASE] Release of intcache 1.1.1 (Nicole Cordes)
2020-02-10 8a5f5b6 [FEATURE] Add functional tests (Nicole Cordes)
2020-02-10 df593e2 [TASK] Update supporting files to EXT:my_skeleton 0.3.0 (Nicole Cordes)
2020-02-06 513e6b6 Allow disabling intcache for typoscript page type (mueller-sebastian)

2019-09-03         [RELEASE] Release of intcache 1.1.0 (Nicole Cordes)
2019-09-03 9093088 [FEATURE] Add TypoScript configuration to disable intcache handling (Nicole Cordes)

2019-09-03         [RELEASE] Release of intcache 1.0.2 (Nicole Cordes)
2019-09-03 c4af82d [BUGFIX] Correctly enable addQueryString in typolink configuration (Nicole Cordes)
2019-09-03 8897f4a [BUGFIX] Allow overriding cache configuration (Nicole Cordes)

2019-09-03         [RELEASE] Release of intcache 1.0.1 (Nicole Cordes)
2019-09-03 ae2e968 [BUGFIX] Enforce separate page cache entries (Nicole Cordes)
2019-09-03 fbea474 [BUGFIX] Use proper pageNotFound handling (Nicole Cordes)

2019-08-21         [RELEASE] Release of intcache 1.0.0 (Nicole Cordes)
2019-08-21 9ddae16 [TASK] Update README.md (Nicole Cordes)
2019-08-21 db5e87d [BUGFIX] Use absRefPrefix instead of absolute url (Nicole Cordes)
2019-06-16 7f1f8df [BUGFIX] Stick Travis CI dist to trusty (Nicole Cordes)
2019-06-06 cee9506 [TASK] Update supporting files to EXT:my_skeleton 0.2.1 (Nicole Cordes)
2019-06-06 0dbb4e9 [FEATURE] Add content rendering in own page type (Nicole Cordes)
2019-06-06 00f5428 [FEATURE] Add hook to turn INT objects into link urls (Nicole Cordes)
2019-06-05 774e98e [FEATURE] Add TypoScript and Fluid templates (Nicole Cordes)
2019-06-03 961b631 [TASK] Add basic extension files (Nicole Cordes)


Tests/Functional/AbstractFunctionalTestCase.php
<?php
namespace IchHabRecht\Intcache\Tests\Functional;

/*
 * This file is part of the TYPO3 extension intcache.
 *
 * (c) Nicole Cordes <typo3@cordes.co>
 * It originated from the EXT:vcc package (https://packagist.org/packages/cpsit/vcc)
 *
 * It is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License, either version 2
 * of the License, or any later version.
 *
 * For the full copyright and license information, please read the
 * LICENSE file that was distributed with this source code.
 */

use Nimut\TestingFramework\Http\Response;
use Nimut\TestingFramework\TestCase\FunctionalTestCase;
use PHPUnit\Util\PHP\DefaultPhpProcess;

abstract class AbstractFunctionalTestCase extends FunctionalTestCase
{
    /**
     * @var array
     */
    protected $testExtensionsToLoad = [
        'typo3conf/ext/intcache',
    ];

    protected function setUp()
    {
        parent::setUp();

        $this->importDataSet('ntf://Database/pages.xml');
        $this->setUpFrontendRootPage(
            1,
            [
                'EXT:intcache/Configuration/TypoScript/setup.txt',
                'EXT:intcache/Tests/Functional/Fixtures/TypoScript/intcache.ts',
            ]
        );
    }

    protected function getFrontendResponseWithQuery($pageId, $languageId = 0, $backendUserId = 0, $workspaceId = 0, $failOnFailure = true, $frontendUserId = 0, $additionalParameter = '')
    {
        $pageId = (int)$pageId;
        $languageId = (int)$languageId;

        if (!empty($frontendUserId)) {
            $additionalParameter .= '&frontendUserId=' . (int)$frontendUserId;
        }
        if (!empty($backendUserId)) {
            $additionalParameter .= '&backendUserId=' . (int)$backendUserId;
        }
        if (!empty($workspaceId)) {
            $additionalParameter .= '&workspaceId=' . (int)$workspaceId;
        }

        $arguments = [
            'documentRoot' => $this->getInstancePath(),
            'requestUrl' => 'http://localhost/?id=' . $pageId . '&L=' . $languageId . $additionalParameter,
        ];

        $template = new \Text_Template('ntf://Frontend/Request.tpl');
        $template->setVar(
            [
                'arguments' => var_export($arguments, true),
                'originalRoot' => ORIGINAL_ROOT,
                'ntfRoot' => __DIR__ . '/../../',
            ]
        );

        $php = DefaultPhpProcess::factory();
        $response = $php->runJob($template->render());
        $result = json_decode($response['stdout'], true);

        if ($result === null) {
            $this->fail('Frontend Response is empty.' . LF . 'Error: ' . LF . $response['stderr']);
        }

        if ($failOnFailure && $result['status'] === Response::STATUS_Failure) {
            $this->fail('Frontend Response has failure:' . LF . $result['error']);
        }

        $response = new Response($result['status'], $result['content'], $result['error']);

        return $response;
    }
}


Tests/Functional/Fixtures/TypoScript/intcache.ts
config {
    disableAllHeaderCode = 1
}

lib.intcache.settings.typeNum = 4242
intcache.typeNum = 4242

page = PAGE
page {
    10 = COA_INT
    10 {
        10 = TEXT
        10.value = Hello world!
    }
}

page2 < page
page2 {
    typeNum = 2

    config.intcache = 0
}


Tests/Functional/Hooks/IntcacheEnabledTestCase.php
<?php
namespace IchHabRecht\Intcache\Tests\Functional\Hooks;

/*
 * This file is part of the TYPO3 extension intcache.
 *
 * (c) Nicole Cordes <typo3@cordes.co>
 * It originated from the EXT:vcc package (https://packagist.org/packages/cpsit/vcc)
 *
 * It is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License, either version 2
 * of the License, or any later version.
 *
 * For the full copyright and license information, please read the
 * LICENSE file that was distributed with this source code.
 */

use IchHabRecht\Intcache\Tests\Functional\AbstractFunctionalTestCase;

class IntcacheEnabledTestCase extends AbstractFunctionalTestCase
{
    /**
     * @test
     */
    public function intcacheIsEnabled()
    {
        $response = $this->getFrontendResponse(1);

        $this->assertSame('success', $response->getStatus());

        $content = $response->getContent();

        $this->assertContains('class="intcache intcache-item intcache-link"', $content);

        $matches = null;
        preg_match('/data-src="([^"]+)"/', $content, $matches);

        $this->assertArrayHasKey(1, $matches);

        $urlParams = parse_url($matches[1]);
        $this->assertArrayHasKey('query', $urlParams);

        $response = $this->getFrontendResponseWithQuery(1, 0, 0, 0, true, 0, '&' . $urlParams['query']);

        $this->assertSame('success', $response->getStatus());
        $this->assertSame('Hello world!', $response->getContent());
    }
}


ext_localconf.php
<?php
defined('TYPO3_MODE') || die();

call_user_func(function () {
    if (!is_array($GLOBALS['TYPO3_CONF_VARS']['SYS']['caching']['cacheConfigurations']['tx_intcache_int'])) {
        $GLOBALS['TYPO3_CONF_VARS']['SYS']['caching']['cacheConfigurations']['tx_intcache_int'] = [];
    }
    $GLOBALS['TYPO3_CONF_VARS']['SYS']['caching']['cacheConfigurations']['tx_intcache_int'] = array_merge(
        [
            'frontend' => \TYPO3\CMS\Core\Cache\Frontend\VariableFrontend::class,
            'backend' => \TYPO3\CMS\Core\Cache\Backend\Typo3DatabaseBackend::class,
            'options' => [
                'compression' => true,
                'defaultLifetime' => 864000, // 10 days
            ],
            'groups' => ['system'],
        ],
        $GLOBALS['TYPO3_CONF_VARS']['SYS']['caching']['cacheConfigurations']['tx_intcache_int']
    );

    $GLOBALS['TYPO3_CONF_VARS']['SC_OPTIONS']['tslib/class.tslib_fe.php']['createHashBase']['intcache'] =
        \IchHabRecht\Intcache\Hooks\CreateHashBaseHook::class . '->createHashBase';
    $GLOBALS['TYPO3_CONF_VARS']['SC_OPTIONS']['tslib/class.tslib_fe.php']['contentPostProc-all']['intcache'] =
        \IchHabRecht\Intcache\Hooks\ContentPostProcessHook::class . '->replaceIntScripts';
});


Classes/Renderer/IntcacheRenderer.php
<?php
namespace IchHabRecht\Intcache\Renderer;

/*
 * This file is part of the TYPO3 extension intcache.
 *
 * (c) Nicole Cordes <typo3@cordes.co>
 * It originated from the EXT:vcc package (https://packagist.org/packages/cpsit/vcc)
 *
 * It is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License, either version 2
 * of the License, or any later version.
 *
 * For the full copyright and license information, please read the
 * LICENSE file that was distributed with this source code.
 */

use IchHabRecht\Intcache\Exception\Exception;
use TYPO3\CMS\Core\Cache\CacheManager;
use TYPO3\CMS\Core\Cache\Frontend\FrontendInterface;
use TYPO3\CMS\Core\Http\ImmediateResponseException;
use TYPO3\CMS\Core\Utility\GeneralUtility;
use TYPO3\CMS\Extbase\Security\Cryptography\HashService;
use TYPO3\CMS\Frontend\Controller\ErrorController;
use TYPO3\CMS\Frontend\Controller\TypoScriptFrontendController;
use TYPO3\CMS\Frontend\Page\PageAccessFailureReasons;

class IntcacheRenderer
{
    /**
     * @var HashService
     */
    protected $hashService;

    /**
     * @var FrontendInterface
     */
    protected $intCache;

    /**
     * @var IntObjectRenderer
     */
    protected $intObjectRenderer;

    /**
     * @var TypoScriptFrontendController
     */
    protected $typoScriptFrontendController;

    public function __construct(
        HashService $hashService = null,
        FrontendInterface $intCache = null,
        IntObjectRenderer $intObjectRenderer = null,
        TypoScriptFrontendController $typoScriptFrontendController = null
    ) {
        $this->hashService = $hashService ?: GeneralUtility::makeInstance(HashService::class);
        $this->intCache = $intCache ?: GeneralUtility::makeInstance(CacheManager::class)->getCache('tx_intcache_int');
        $this->intObjectRenderer = $intObjectRenderer ?: GeneralUtility::makeInstance(IntObjectRenderer::class);
        $this->typoScriptFrontendController = $typoScriptFrontendController ?: $GLOBALS['TSFE'];
    }

    public function render()
    {
        $arguments = GeneralUtility::_GET('tx_intcache');
        if (empty($arguments['identifier']) || !is_string($arguments['identifier'])) {
            throw new  Exception('Missing identifier', 1559825922);
        }

        $cacheIdentifier = $this->hashService->validateAndStripHmac($arguments['identifier']);

        $configuration = $this->intCache->get($cacheIdentifier);

        if (empty($configuration)) {
            $this->setPageNotFound();
        }

        if (!empty($configuration['conf']['cache_timeout'])) {
            $this->typoScriptFrontendController->page['cache_timeout'] = $configuration['conf']['cache_timeout'];
        } else {
            $this->typoScriptFrontendController->set_no_cache('Intcache response is non-cacheable by default');
        }

        return $this->intObjectRenderer->render($configuration);
    }

    protected function setPageNotFound()
    {
        if (class_exists('TYPO3\\CMS\\Core\\Http\\ImmediateResponseException')) {
            $response = GeneralUtility::makeInstance(ErrorController::class)->pageNotFoundAction(
                $GLOBALS['TYPO3_REQUEST'],
                'No intcache configuration found',
                [
                    'code' => PageAccessFailureReasons::PAGE_NOT_FOUND,
                ]
            );
            throw new ImmediateResponseException($response, 1533931329);
        }

        $this->typoScriptFrontendController->pageNotFoundAndExit(
            'No intcache configuration found'
        );
    }
}


Classes/Renderer/IntObjectRenderer.php
<?php
namespace IchHabRecht\Intcache\Renderer;

/*
 * This file is part of the TYPO3 extension intcache.
 *
 * (c) Nicole Cordes <typo3@cordes.co>
 * It originated from the EXT:vcc package (https://packagist.org/packages/cpsit/vcc)
 *
 * It is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License, either version 2
 * of the License, or any later version.
 *
 * For the full copyright and license information, please read the
 * LICENSE file that was distributed with this source code.
 */

use TYPO3\CMS\Core\Page\PageRenderer;
use TYPO3\CMS\Core\Utility\GeneralUtility;
use TYPO3\CMS\Frontend\ContentObject\ContentObjectRenderer;

class IntObjectRenderer
{
    /**
     * @param array $configuration
     * @return string
     */
    public function render(array $configuration)
    {
        if (empty($configuration['cObj']) || empty($configuration['type'])) {
            return '';
        }

        $contentObjectRenderer = unserialize($configuration['cObj']);
        if (!$contentObjectRenderer instanceof ContentObjectRenderer) {
            return '';
        }

        $content = '';
        switch ($configuration['type']) {
            case 'COA':
                $content = $contentObjectRenderer->cObjGetSingle('COA', $configuration['conf']);
                break;
            case 'FUNC':
                $content = $contentObjectRenderer->cObjGetSingle('USER', $configuration['conf']);
                break;
            case 'POSTUSERFUNC':
                $content = $contentObjectRenderer->callUserFunction($configuration['postUserFunc'], $configuration['conf'], $configuration['content']);
                break;
        }

        $pageRenderer = GeneralUtility::makeInstance(PageRenderer::class);
        $pageRenderer->setTemplateFile('EXT:intcache/Resources/Private/Templates/PageRenderer.html');
        $page = $pageRenderer->render(PageRenderer::PART_HEADER);
        $page .= $content;
        // The PageRenderer gets reset after render, so the template must be reassigned
        $pageRenderer->setTemplateFile('EXT:intcache/Resources/Private/Templates/PageRenderer.html');
        $page .= $pageRenderer->render(PageRenderer::PART_FOOTER);

        return $page;
    }
}


Classes/Hooks/ContentPostProcessHook.php
<?php
namespace IchHabRecht\Intcache\Hooks;

/*
 * This file is part of the TYPO3 extension intcache.
 *
 * (c) Nicole Cordes <typo3@cordes.co>
 * It originated from the EXT:vcc package (https://packagist.org/packages/cpsit/vcc)
 *
 * It is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License, either version 2
 * of the License, or any later version.
 *
 * For the full copyright and license information, please read the
 * LICENSE file that was distributed with this source code.
 */

use IchHabRecht\Intcache\Exception\Exception;
use IchHabRecht\Intcache\Renderer\IntObjectRenderer;
use Psr\Http\Message\ServerRequestInterface;
use TYPO3\CMS\Core\Cache\CacheManager;
use TYPO3\CMS\Core\Cache\Frontend\FrontendInterface;
use TYPO3\CMS\Core\Core\ApplicationContext;
use TYPO3\CMS\Core\Routing\PageArguments;
use TYPO3\CMS\Core\Utility\GeneralUtility;
use TYPO3\CMS\Extbase\Security\Cryptography\HashService;
use TYPO3\CMS\Frontend\ContentObject\ContentObjectRenderer;
use TYPO3\CMS\Frontend\Controller\TypoScriptFrontendController;

class ContentPostProcessHook
{
    /**
     * @var ApplicationContext
     */
    protected $applicationContext;

    /**
     * @var HashService
     */
    protected $hashService;

    /**
     * @var FrontendInterface
     */
    protected $intCache;

    /**
     * @var IntObjectRenderer
     */
    protected $intObjectRenderer;

    /**
     * @var TypoScriptFrontendController
     */
    protected $typoScriptFrontendController;

    public function __construct(
        ApplicationContext $applicationContext = null,
        HashService $hashService = null,
        FrontendInterface $intCache = null,
        IntObjectRenderer $intObjectRenderer = null
    ) {
        $this->applicationContext = $applicationContext ?: GeneralUtility::getApplicationContext();
        $this->hashService = $hashService ?: GeneralUtility::makeInstance(HashService::class);
        $this->intCache = $intCache ?: GeneralUtility::makeInstance(CacheManager::class)->getCache('tx_intcache_int');
        $this->intObjectRenderer = $intObjectRenderer ?: GeneralUtility::makeInstance(IntObjectRenderer::class);
    }

    public function replaceIntScripts(array $parameter)
    {
        $this->typoScriptFrontendController = $parameter['pObj'];
        if ($_SERVER['REQUEST_METHOD'] === 'POST'
            || empty($this->typoScriptFrontendController->config['INTincScript'])
            || empty($this->typoScriptFrontendController->config['config']['intcache'])
        ) {
            return;
        }

        if (empty($this->typoScriptFrontendController->tmpl->setup['lib.']['intcache.']['settings.']['typeNum'])) {
            throw new Exception('Page TypeNum for INT rendering must be set', 1559647300);
        }

        $cacheTag = 'newHash_' . $this->typoScriptFrontendController->newHash;
        $this->intCache->flushByTag($cacheTag);

        $useQueryString = !empty($this->typoScriptFrontendController->cHash);
        $request = $GLOBALS['TYPO3_REQUEST'] ?? null;
        if (!$useQueryString && $request instanceof ServerRequestInterface) {
            $pageArguments = $request->getAttribute('routing');
            if ($pageArguments instanceof PageArguments) {
                $useQueryString = !empty($pageArguments->getStaticArguments());
            }
        }
        $addQueryStringMethod = $useQueryString ? 'GET' : '';

        $content = $this->typoScriptFrontendController->content;
        foreach ($this->typoScriptFrontendController->config['INTincScript'] as $identifier => $configuration) {
            $matches = [];
            if (preg_match('/<!--\s*' . preg_quote($identifier) . '\s*-->/i', $content, $matches)) {
                $cacheIdentifier = md5(json_encode($configuration) . $cacheTag);
                $this->intCache->set($cacheIdentifier, $configuration, [$cacheTag]);

                $contentObjectRenderer = (!empty($configuration['cObj']))
                    ? unserialize($configuration['cObj'])
                    : $this->typoScriptFrontendController->cObj;
                if (!$contentObjectRenderer instanceof ContentObjectRenderer) {
                    $contentObjectRenderer = $this->typoScriptFrontendController->cObj;
                }

                $intcacheContent = '';
                if ($this->applicationContext->isDevelopment()) {
                    $intcacheContent = $this->intObjectRenderer->render($configuration);
                }

                $intcacheLink = $contentObjectRenderer->typoLink_URL(
                    [
                        'parameter' => implode(',', [
                            $this->typoScriptFrontendController->id,
                            $this->typoScriptFrontendController->tmpl->setup['lib.']['intcache.']['settings.']['typeNum'],
                        ]),
                        'addQueryString' => !empty($addQueryStringMethod),
                        'addQueryString.' => [
                            'method' => $addQueryStringMethod,
                        ],
                        'additionalParams' => '&tx_intcache[identifier]=' . $this->hashService->appendHmac($cacheIdentifier),
                        'useCacheHash' => 1,
                        'wrap' => $this->typoScriptFrontendController->absRefPrefix . '|',
                    ]
                );

                $intcacheIdentifier = $this->typoScriptFrontendController->uniqueHash('intcache');

                $data = array_merge(
                    (array)$contentObjectRenderer->data,
                    [
                        'intcache_content' => $intcacheContent,
                        'intcache_identifier' => $intcacheIdentifier,
                        'intcache_link' => $intcacheLink,
                    ]
                );

                $contentObjectRenderer->start($data, $contentObjectRenderer->getCurrentTable());
                $intContent = $contentObjectRenderer->cObjGetSingle($this->typoScriptFrontendController->tmpl->setup['lib.']['intcache'], $this->typoScriptFrontendController->tmpl->setup['lib.']['intcache.']);
                $content = str_replace($matches[0], $intContent, $content);
            }
            unset($this->typoScriptFrontendController->config['INTincScript'][$identifier]);
        }

        $this->typoScriptFrontendController->content = $content;

        if (empty($this->typoScriptFrontendController->config['INTincScript'])) {
            // Add additional header parts
            $this->typoScriptFrontendController->INTincScript();
            unset($this->typoScriptFrontendController->config['INTincScript']);
        }
    }
}


ext_emconf.php
<?php

/***************************************************************
 * Extension Manager/Repository config file for ext "intcache".
 *
 * Auto generated 24-04-2020 11:04
 *
 * Manual updates:
 * Only the data in the array - everything else is removed by next
 * writing. "version" and "dependencies" must not be touched!
 ***************************************************************/

$EM_CONF[$_EXTKEY] = array (
  'title' => 'intcache',
  'description' => 'Turn uncachable page objects into cacheable links',
  'category' => 'misc',
  'author' => 'Nicole Cordes',
  'author_email' => 'typo3@cordes.co',
  'author_company' => 'biz-design',
  'state' => 'stable',
  'uploadfolder' => 0,
  'createDirs' => '',
  'clearCacheOnLoad' => 0,
  'version' => '1.1.3',
  'constraints' => 
  array (
    'depends' => 
    array (
      'typo3' => '7.6.0-10.4.99',
    ),
    'conflicts' => 
    array (
    ),
    'suggests' => 
    array (
    ),
  ),
  '_md5_values_when_last_written' => 'a:27:{s:9:"ChangeLog";s:4:"1155";s:7:"LICENSE";s:4:"b234";s:9:"README.md";s:4:"b0b3";s:13:"composer.json";s:4:"f285";s:13:"composer.lock";s:4:"1bd0";s:12:"ext_icon.png";s:4:"2f84";s:17:"ext_localconf.php";s:4:"173f";s:16:"phpunit.xml.dist";s:4:"041c";s:24:"sonar-project.properties";s:4:"b13b";s:31:"Classes/Exception/Exception.php";s:4:"8ec8";s:40:"Classes/Hooks/ContentPostProcessHook.php";s:4:"22df";s:36:"Classes/Hooks/CreateHashBaseHook.php";s:4:"ceeb";s:38:"Classes/Renderer/IntObjectRenderer.php";s:4:"519b";s:37:"Classes/Renderer/IntcacheRenderer.php";s:4:"6214";s:42:"Configuration/TCA/Overrides/tt_content.php";s:4:"51da";s:38:"Configuration/TypoScript/constants.txt";s:4:"e2f4";s:34:"Configuration/TypoScript/setup.txt";s:4:"8728";s:43:"Resources/Private/Language/locallang_be.xlf";s:4:"6677";s:41:"Resources/Private/Templates/Intcache.ajax";s:4:"8c84";s:40:"Resources/Private/Templates/Intcache.esi";s:4:"c1cd";s:40:"Resources/Private/Templates/Intcache.ssi";s:4:"0e9b";s:45:"Resources/Private/Templates/PageRenderer.html";s:4:"9153";s:36:"Resources/Public/Icons/Extension.svg";s:4:"8ec3";s:47:"Tests/Functional/AbstractFunctionalTestCase.php";s:4:"d31c";s:48:"Tests/Functional/Fixtures/TypoScript/intcache.ts";s:4:"8467";s:51:"Tests/Functional/Hooks/IntcacheDisabledTestCase.php";s:4:"2ba1";s:50:"Tests/Functional/Hooks/IntcacheEnabledTestCase.php";s:4:"bbdf";}',
);

